test/integration/formatting_stack/strategies.clj

```diff
@@ -0,0 +1,153 @@
+(ns integration.formatting-stack.strategies
+  (:require
+   [clojure.java.shell :refer [sh]]
+   [clojure.test :refer [deftest is testing]]
+   [formatting-stack.strategies :as sut])
+  (:import
+   (java.io File)))
+
+(def ^File deletable-file (File. "test-resources/deletable.clj"))
+
+(def ^String createable-filename "test-resources/createable.clj")
+
+(def creatable-contents (pr-str '(ns foo)))
+
+(assert (-> deletable-file .exists))
+
+(assert (not (-> createable-filename File. .exists)))
+
+(defn assert-pristine-git-status! []
+  (assert (-> (sh "git" "status" "--porcelain")
+              :out
+              #{""})
+          "This test needs the `git status` to be pristine; else it will fail and/or destroy your WIP."))
+
+(defn expect-sane-output! [corpus]
+  (let [n (count corpus)]
+    (is (pos? n))
+    (doseq [^String filename corpus
+            :let [f (File. filename)]]
+      (is (-> f .exists))
+      (testing "It emits absolutized filenames"
+        (is (= filename
+               (-> f .getCanonicalPath)))))))
+
+(deftest all-files
+
+  (assert-pristine-git-status!)
+
+  (testing "It runs without errors, exercising its specs, even in face of deleted (but not staged) files"
+    (try
+      (-> deletable-file .delete)
+      (expect-sane-output! (sut/all-files :files []))
+      (finally
+        (sh "git" "checkout" (str deletable-file))
+        (assert (-> deletable-file .exists)))))
+
+  (assert-pristine-git-status!)
+
+  (testing "It runs without errors, exercising its specs, even in face of files staged for deletion"
+    (try
+      (-> deletable-file .delete)
+      (sh "git" "add" "-A")
+      (expect-sane-output! (sut/all-files :files []))
+      (finally
+        (sh "git" "reset" "--" (str deletable-file))
+        (sh "git" "checkout" (str deletable-file))
+        (assert (-> deletable-file .exists))))))
+
+(deftest git-diff-against-default-branch
+
+  (assert-pristine-git-status!)
+
+  (testing "It runs without errors, exercising its specs, even in face of deleted (but not staged) files"
+    (let [;; f-stack's first ever commit. This ensures a large, diverse corpus:
+          first-commit "ebb0ca8"]
```

Comment:
could be [`git rev-list --max-parents=0 HEAD`](https://stackoverflow.com/questions/5188914/how-to-show-first-commit-by-git-log). to keep the actual commit hash out of the source

Reply:
The let can be one level higher, (or a delayed def with the command above) preventing duplication on L77

Reply:
> git rev-list --max-parents=0 HEAD.

Will try